### PR TITLE
Refactor ConsoleProcessList

### PR DESCRIPTION
**Pull request by lhecker**
This commit is just a slight refactor of `ConsoleProcessList` which I've noticed
was in a poor shape. It replaces iterators with for-range loops, etc.

Additionally this fixes a bug in `SetConsoleWindowOwner`, where it used to
default to the newest client process instead of the oldest.

Finally, it changes the process container type from a doubly linked list
over to a simple array/vector, because using linked lists for heap allocated
elements felt quite a bit silly. To preserve the previous behavior of
`GetProcessList`, it simply iterates through the vector backwards.

## Validation Steps Performed
* All unit/feature tests pass ✅
* Launching a TUI application inside pwsh inside cmd
  and exiting kills all 3 applications ✅

### CODE REVIEW

File: src/interactivity/win32/windowio.cpp [hunk 1]
```diff
@@ -78,7 +78,7 @@ VOID SetConsoleWindowOwner(const HWND hwnd, _Inout_opt_ ConsoleProcessHandle* pP
     else
     {
         // Find a process to own the console window. If there are none then let's use conhost's.
-        pProcessData = gci.ProcessHandleList.GetFirstProcess();
+        pProcessData = gci.ProcessHandleList.GetOldestProcess();
```
Comment (lhecker) [inline src/interactivity/win32/windowio.cpp:81]: This is a behavior change: `GetFirstProcess()` used to return the newest process, but I think this was wrong... conhost's PID was the last (= oldest) entry if anything. So I've chosen to rename the function to better convey its meaning.

File: src/server/ProcessList.cpp [hunk 2]
```diff
@@ -30,51 +27,27 @@ using namespace Microsoft::Console::Interactivity;
 [[nodiscard]] HRESULT ConsoleProcessList::AllocProcessData(const DWORD dwProcessId,
                                                            const DWORD dwThreadId,
                                                            const ULONG ulProcessGroupId,
-                                                           _In_opt_ ConsoleProcessHandle* const pParentProcessData,
```
Comment (lhecker) [inline src/server/ProcessList.cpp:33]: `pParentProcessData` wasn't used for anything except to change the return value.

File: src/server/ProcessList.cpp [hunk 3]
```diff
@@ -30,51 +27,27 @@ using namespace Microsoft::Console::Interactivity;
...
-    if (nullptr != pProcessData)
+    if (!pProcessData)
     {
-        // In the GenerateConsoleCtrlEvent it's OK for this process to already have a ProcessData object. However, the other case is someone
-        // connecting to our LPC port and they should only do that once, so we fail subsequent connection attempts.
-        if (nullptr == pParentProcessData)
-        {
-            return E_FAIL;
-            // TODO: MSFT: 9574803 - This fires all the time. Did it always do that?
-            //RETURN_HR(E_FAIL);
```
Comment (lhecker) [inline src/server/ProcessList.cpp:47]: TODO resolved: There's only two callers. `IoDispatchers::ConsoleHandleConnectionRequest` calls this method with a `nullptr` for `pParentProcessData`.

The other caller is `ApiDispatchers::ServerGenerateConsoleCtrlEvent`, which already has a `RETURN_HR_IF_NULL(E_INVALIDARG, pParentProcessData)`, preventing this method from ever being called in the first place.

Comment (lhecker) [reply src/server/ProcessList.cpp:47]: ...unless of course `ConsoleHandleConnectionRequest`'s intention is to guard against accepting a connection twice? Is that something we need to guard against?

File: src/server/ProcessList.cpp [hunk 4]
```diff
@@ -30,51 +27,29 @@ using namespace Microsoft::Console::Interactivity;
...
-    if (nullptr != pProcessData)
+    if (pProcessData)
     {
-        // In the GenerateConsoleCtrlEvent it's OK for this process to already have a ProcessData object. However, the other case is someone
-        // connecting to our LPC port and they should only do that once, so we fail subsequent connection attempts.
-        if (nullptr == pParentProcessData)
-        {
-            return E_FAIL;
-            // TODO: MSFT: 9574803 - This fires all the time. Did it always do that?
-            //RETURN_HR(E_FAIL);
```
Comment (lhecker) [inline src/server/ProcessList.cpp:47]: There's only two callers. `ApiDispatchers::ServerGenerateConsoleCtrlEvent` always calls this with a valid `pParentProcessData` and won't ever run into this condition. `IoDispatchers::ConsoleHandleConnectionRequest` calls this method with a `nullptr` instead. This indicates that this error triggers, because we're getting redundant connection requests for the same process ID.

The new code will now always return `E_FAIL`, because the former, `ApiDispatchers::ServerGenerateConsoleCtrlEvent`, actually only calls this function if the `dwProcessId` is not in the list:
https://github.com/microsoft/terminal/blob/437b5ac595915830194498eca8825e6037e288c0/src/server/ApiDispatchersInternal.cpp#L83-L84

File: src/host/input.cpp [hunk 5]
```diff
@@ -275,27 +275,17 @@ void ProcessCtrlEvents()
...
-    if (FAILED(hr) || cProcessHandleList == 0)
+    if (FAILED(hr) || termRecords.empty())
     {
         gci.UnlockConsole();
         return;
     }
 
     // Copy ctrl flags.
-    auto CtrlFlags = gci.CtrlFlags;
-    FAIL_FAST_IF(!(!((CtrlFlags & (CONSOLE_CTRL_CLOSE_FLAG | CONSOLE_CTRL_BREAK_FLAG | CONSOLE_CTRL_C_FLAG)) && (CtrlFlags & (CONSOLE_CTRL_LOGOFF_FLAG | CONSOLE_CTRL_SHUTDOWN_FLAG)))));
```
Comment (lhecker) [inline src/host/input.cpp:296]: There's no good reason for this to fail-fast.

File: src/server/ApiDispatchersInternal.cpp [hunk 6]
```diff
@@ -90,11 +90,14 @@ using Microsoft::Console::Interactivity::ServiceLocator;
...
-                                                                    ProcessHandle,
-                                                                    nullptr));
+
+            const auto hr = gci.ProcessHandleList.AllocProcessData(a->ProcessGroupId, 0, a->ProcessGroupId, nullptr);
+            // AllocProcessData returns E_FAIL if the given process ID already exists.
+            // In that case we don't really need to return an error and can just carry on.
+            if (FAILED(hr) && hr != E_FAIL)
+            {
+                RETURN_HR(hr);
+            }
```
Comment (lhecker) [inline src/server/ApiDispatchersInternal.cpp:100]: Now that `AllocProcessData` consistently returns `E_FAIL` if the process ID already exists, we need this bit of extra code here. TBH I don't quite understand why this adds a "ProcessGroupId" as a process ID to the process list here...

File: src/interactivity/win32/windowio.cpp (hunk 1 above)
Comment (lhecker) [reply src/interactivity/win32/windowio.cpp:81]: > [any old](https://en.wiktionary.org/wiki/any_old) has an idiomatic meaning, though.

Ah good point.

> Is the comment "_If there are none then let's use conhost's._" still accurate, or does this now use conhost's process even if others exist?

I'm not sure it ever used to be accurate. `conhost` itself isn't part of the process lists - console clients are (i.e. shells and other console subsystem applications). I wouldn't know how conhost could be its own client.
In reality this code path is only invoked via [`Window::SetOwner`](https://github.com/microsoft/terminal/blob/437b5ac595915830194498eca8825e6037e288c0/src/host/srvinit.cpp#L282) when the "root process" exits (the first process that is launched - usually your shell, like `cmd.exe` for instance), but there are still other clients attached. This is for instance reproducible if you run `pwsh.exe` inside `cmd.exe` and close the window. Both processes will get get a close event simultaneously, but PowerShell is slow and so it'll take longer to exit then `cmd.exe`. Once `cmd.exe` is gone, `conhost` will need a new "root process" and in this case that would be `pwsh.exe` (until it also exits a hundred or so milliseconds later).

So basically, the comment is at least generally wrong/inaccurate, but I don't want to touch it, because it might prove to be a useful nugget of information if looked up via `git blame` in the future. The old code used to use the newest process as the new root process and that seems like a mistake to me. If anything it should be the oldest one as far as I can see.

File: src/server/ProcessList.cpp [hunk 7]
```diff
@@ -210,67 +186,48 @@ ConsoleProcessHandle* ConsoleProcessList::FindProcessByGroupId(_In_ ULONG ulProc
 // - E_OUTOFMEMORY in a low memory situation.
 [[nodiscard]] HRESULT ConsoleProcessList::GetTerminationRecordsByGroupId(const DWORD dwLimitingProcessId,
                                                                          const bool fCtrlClose,
-                                                                         _Outptr_result_buffer_all_(*pcRecords) ConsoleProcessTerminationRecord** prgRecords,
-                                                                         _Out_ size_t* const pcRecords) const
+                                                                         std::vector<ConsoleProcessTerminationRecord>& termRecords) const
```
Comment (carlos-zamora) [inline src/server/ProcessList.cpp:189]:
```suggestion
                                                                         _Out_ std::vector<ConsoleProcessTerminationRecord>& termRecords) const
```
nit: if you want to match the SAL annotations

File: src/host/input.cpp [hunk 8]
```diff
@@ -347,20 +344,13 @@ void ProcessCtrlEvents()
...
+            if (!r.hProcess)
             {
                 Status = STATUS_SUCCESS;
             }
         }
-
-        if (rgProcessHandleList[i].hProcess != nullptr)
-        {
-            CloseHandle(rgProcessHandleList[i].hProcess);
-        }
```
Comment (carlos-zamora) [inline src/host/input.cpp:362]: Don't we still need this in the event that `if (NT_SUCCESS(Status))` fails?

File: src/server/ProcessList.cpp (hunk 4 above)
Comment (lhecker) [reply src/server/ProcessList.cpp:47]: No it does get called by `IoDispatchers::ConsoleHandleConnectionRequest`. It sets `pParentProcessData` to `nullptr` which then invokes this branch. It does that so that `AllocProcessData` returns `S_OK` instead of `E_FAIL` if the given process ID already exists.

...but that practically makes the `pParentProcessData` parameter a `bool allowDuplicateProcessId` and I didn't particularly like that idea, when the caller might as well just filter the `E_FAIL` message themselves. This way the `AllocProcessData` is now way way simpler: It either sucessfully inserts a new process into the list or it fails. Simple as that! I felt like that would be a good idea.

File: src/host/input.cpp (hunk 8 above)
Comment (lhecker) [reply src/host/input.cpp:362]: I made it so that the `hProcess` is a `wil::unique_handle` so it'll get cleaned up automatically.

**Comment by DHowett** [comment]
Can I add a validation step?

Launch cmd. Launch `code` inside it. `exit` in CMD once you get your prompt back.

Who owns the window? My understanding of the code before your rewrite was that `code.exe` _should_. I'm not seeing that in practice however. _Huh,_

File: src/server/ProcessList.cpp [hunk 9]
```diff
@@ -30,51 +27,29 @@ using namespace Microsoft::Console::Interactivity;
...
 
     try
     {
-        pProcessData = new ConsoleProcessHandle(dwProcessId,
-                                                dwThreadId,
-                                                ulProcessGroupId);
-
-        // Some applications, when reading the process list through the GetConsoleProcessList API, are expecting
-        // the returned list of attached process IDs to be from newest to oldest.
-        // As such, we have to put the newest process into the head of the list.
```
Comment (zadjii-msft) [inline src/server/ProcessList.cpp:67]: This was `LOAD BEARING` in the past. What happened to this?

Comment (zadjii-msft) [reply src/server/ProcessList.cpp:67]: answer: this moved to `GetProcessList`

**Review by zadjii-msft — APPROVED** [review]

**Review by DHowett — CHANGES_REQUESTED** [review]
(my signoff is contingent on making sure the `cmd` => `code` => `exit` case works at least close to properly!)

**Comment by lhecker** [comment]
@DHowett Did you have a specific invocation of VS Code in mind? The `code` executable only spawns the electron application and then exits, so we receive a disconnect event and `ConsoleProcessList::FreeProcessData` is called (I just confirmed it in a debugger).

File: src/interactivity/win32/windowio.cpp [hunk 10]
```diff
@@ -78,11 +78,11 @@ VOID SetConsoleWindowOwner(const HWND hwnd, _Inout_opt_ ConsoleProcessHandle* pP
     else
     {
         // Find a process to own the console window. If there are none then let's use conhost's.
-        pProcessData = gci.ProcessHandleList.FindProcessInList(ConsoleProcessList::ROOT_PROCESS_ID);
+        pProcessData = gci.ProcessHandleList.GetRootProcess();
         if (!pProcessData)
         {
             // No root process ID? Pick the oldest existing process.
-            pProcessData = gci.ProcessHandleList.GetFirstProcess();
+            pProcessData = gci.ProcessHandleList.GetOldestProcess();
```
Comment (DHowett) [inline src/interactivity/win32/windowio.cpp:85]: thanks

File: src/server/ApiDispatchersInternal.cpp [hunk 11]
```diff
@@ -90,11 +90,17 @@ using Microsoft::Console::Interactivity::ServiceLocator;
...
-                                                                    a->ProcessGroupId,
-                                                                    ProcessHandle,
-                                                                    nullptr));
+
+            const auto hr = gci.ProcessHandleList.AllocProcessData(a->ProcessGroupId,
+                                                                   0,
+                                                                   a->ProcessGroupId,
+                                                                   nullptr);
+            // AllocProcessData returns E_FAIL if the given process ID already exists.
+            // The process ID already existing isn't really an error though so we ignore it.
```
Comment (DHowett) [inline src/server/ApiDispatchersInternal.cpp:99]: you could use `S_FALSE` for a non-error case if want

File: src/server/ProcessList.h [hunk 12]
```diff
@@ -23,33 +23,33 @@ Revision History:
...
+    ConsoleProcessHandle* GetRootProcess() const;
 
     [[nodiscard]] HRESULT GetTerminationRecordsByGroupId(const DWORD dwLimitingProcessId,
                                                          const bool fCtrlClose,
-                                                         _Outptr_result_buffer_all_(*pcRecords) ConsoleProcessTerminationRecord** prgRecords,
-                                                         _Out_ size_t* const pcRecords) const;
+                                                         std::vector<ConsoleProcessTerminationRecord>& termRecords) const;
 
-    ConsoleProcessHandle* GetFirstProcess() const;
+    ConsoleProcessHandle* GetOldestProcess() const;
```
Comment (DHowett) [inline src/server/ProcessList.h:52]: Hmm. Mind relocating these to be near eachother while you're here?

File: src/server/ProcessList.cpp [hunk 13]
```diff
@@ -30,51 +27,29 @@ using namespace Microsoft::Console::Interactivity;
...
-        _processes.push_front(pProcessData);
-
-        if (nullptr != ppProcessData)
-        {
-            *ppProcessData = pProcessData;
-        }
+        pProcessData = new ConsoleProcessHandle{ dwProcessId, dwThreadId, ulProcessGroupId };
+        _processes.emplace_back(pProcessData);
```
Comment (DHowett) [inline src/server/ProcessList.cpp:43]: nit: if this throws, we leak one ConsoleProcessHandle.

We should capture the return value from `emplace_back` and use that to populate `ppProcessData`

File: src/server/ProcessList.cpp [hunk 14]
```diff
@@ -30,51 +27,29 @@ using namespace Microsoft::Console::Interactivity;
...
     }
     CATCH_RETURN();
 
-    RETURN_HR(S_OK);
+    if (ppProcessData)
```
Comment (DHowett) [inline src/server/ProcessList.cpp:47]: for fun, you could use `wil::assign_to_opt_param`!

File: src/server/ProcessList.cpp [hunk 15]
```diff
@@ -85,47 +60,39 @@ using namespace Microsoft::Console::Interactivity;
...
 
 // Routine Description:
 // - Locates a process handle in this list.
-// - NOTE: Calling FindProcessInList(0) means you want the root process.
 // Arguments:
-// - dwProcessId - ID of the process to search for or ROOT_PROCESS_ID to find the root process.
+// - dwProcessId - ID of the process to search for.
 // Return Value:
 // - Pointer to the process handle information or nullptr if no match was found.
 ConsoleProcessHandle* ConsoleProcessList::FindProcessInList(const DWORD dwProcessId) const
```
Comment (DHowett) [inline src/server/ProcessList.cpp:86]: you've validated that nobody calls `Find(ROOT)` even unintentionally now?

**Review by DHowett — CHANGES_REQUESTED** [review]
Minor correctness issues, but I love it nonetheless

File: src/server/ProcessList.cpp (hunk 15 above)
Comment (lhecker) [reply src/server/ProcessList.cpp:86]: Yeah, there's only a single place calling this function with non-ROOT right now... It's `ApiDispatchers::ServerGenerateConsoleCtrlEvent` which calls it with valid process IDs.

File: src/server/ProcessList.cpp (hunk 13 above)
Comment (lhecker) [reply src/server/ProcessList.cpp:43]: `_processes` is a list of raw pointers actually unfortunately. I'm going to use `std::unique_ptr` in the meantime until we can refactor this code further to not use raw pointers anymore.

Comment (DHowett) [reply src/server/ProcessList.cpp]: Honestly I'm okay with it -- the leak occurs if the emplace fails: `_processes` doesn't take ownership, and we leak it. `unique_ptr` will fix, but just wanted to be clear I'm not as bothered by the list of raw pointers with manual management (though it does gross me out :P)

File: src/server/ProcessList.cpp [hunk 16]
```diff
@@ -31,23 +31,20 @@ using namespace Microsoft::Console::Interactivity;
...
+        pProcessData = std::make_unique<ConsoleProcessHandle>(dwProcessId, dwThreadId, ulProcessGroupId);
+        _processes.emplace_back(pProcessData.get());
     }
     CATCH_RETURN();
 
-    if (ppProcessData)
-    {
-        *ppProcessData = pProcessData;
-    }
+    wil::assign_to_opt_param(ppProcessData, pProcessData.release());
```
Comment (DHowett) [inline src/server/ProcessList.cpp:47]: SCARY! This will delete the processdata if nobody asks for the optional output param

Comment (DHowett) [reply src/server/ProcessList.cpp:47]: wait, no it won't. that's even scarier lol

**Review by DHowett — APPROVED** [review]
